### PR TITLE
Vitis: support different bitstream build strategies

### PR DESCRIPTION
**Pull request by davidbiancolin**
This add some machinery to:
1) Specify optional config overrides to the make system
2) Call out a specific build-strategy by setting a make variable string.

Plumbing this through to the bitbuilder will be a follow up PR.

#### Related PRs / Issues

One part of https://github.com/firesim/firesim/pull/1229 for Vitis. 

#### UI / API Impact

Extra make variables to control the vitis flow. Not exposed to manager yet. 

#### Verilog / AGFI Compatibility

None.

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

**Review by russell-horvath — COMMENTED** [review]
Do we want to keep the naming of the strategy files similar to the F1 specific ones (`strategy_<AREA | TIMING...>`) ? Also I don't know if it makes sense for these small files but storing them in some sort of common directory above `cl_firesim` means we don't have to keep copying them.  Other than that the implementation LGTM.

File: platforms/vitis/cl_firesim/build-bitstream.cfg [hunk 1]
```diff
@@ -1,20 +1,7 @@
...
-
-prop=run.impl_1.STEPS.ROUTE_DESIGN.ARGS.DIRECTIVE=Explore
-prop=run.impl_1.{STEPS.ROUTE_DESIGN.ARGS.MORE OPTIONS}={-tns_cleanup}
-
-prop=run.impl_1.STEPS.PHYS_OPT_DESIGN.IS_ENABLED=true
-prop=run.impl_1.STEPS.PHYS_OPT_DESIGN.ARGS.DIRECTIVE=Explore
-
-prop=run.impl_1.STEPS.POST_ROUTE_PHYS_OPT_DESIGN.IS_ENABLED=true
-prop=run.impl_1.STEPS.POST_ROUTE_PHYS_OPT_DESIGN.ARGS.DIRECTIVE=AggressiveExplore
+# Additional "strategy" configs can be overlayed ontop of this base. See vitis-configs/
```
Comment (davidbiancolin) [inline platforms/vitis/cl_firesim/build-bitstream.cfg:4]:
```suggestion
# Additional "strategy" configs can be overlayed ontop of this base. See build-strategies/
```

**Comment by davidbiancolin** [comment]
I'll make the names consistent, but i'm going to continue copying the files to be consistent with the rest of the collateral.

File: platforms/vitis/cl_firesim/Makefile [hunk 2]
```diff
@@ -118,18 +135,26 @@ ipgen_scripts = $(shell find design/ -iname "*.ipgen.tcl")
...
+build_strategy_path = $(STRATEGIES_DIR)/strategy_$(STRATEGY).cfg
+
+ifneq ($(shell test -f $(build_strategy_path); echo $$?),0)
+$(error Build strategy config at $(build_strategy_path) does not exist)
+endif
+VPP_CONFIG_OVERLAYS += --config $(build_strategy_path)
+
 # link and package the xclbin
-%/$(PROJECT_NAME).link.xclbin: %/$(PROJECT_NAME).xo
+%/$(PROJECT_NAME).link.xclbin: %/$(PROJECT_NAME).xo $(build_strategy_path)
```
Comment (abejgonzalez) [inline platforms/vitis/cl_firesim/Makefile:146]: Shouldn't you also have a requirement be the `CONFIG_OVERLAY`?

File: platforms/vitis/cl_firesim/Makefile [hunk 3]
```diff
@@ -35,13 +35,22 @@ run-sim:   VPP_TARGET_SPECIFIC_ARGS = --target $(TARGET) --config $(buildroot)/s
 # current support on firesim1
 DEVICE ?= xilinx_u250_gen3x16_xdma_3_1_202020_1
 
+# Selects a vitis config, by looking in build-strategies, for an equivalently
+# named cfg file.
+STRATEGY ?= TIMING
+
+# If set, this will very first config file vitis sees.
+CONFIG_OVERLAY ?=
```
Comment (abejgonzalez) [inline platforms/vitis/cl_firesim/Makefile:43]: Why is this needed?

File: platforms/vitis/cl_firesim/Makefile [hunk 4]
```diff
@@ -118,18 +135,26 @@ ipgen_scripts = $(shell find design/ -iname "*.ipgen.tcl")
...
+endif
+VPP_CONFIG_OVERLAYS += --config $(build_strategy_path)
+
 # link and package the xclbin
-%/$(PROJECT_NAME).link.xclbin: %/$(PROJECT_NAME).xo
+%/$(PROJECT_NAME).link.xclbin: %/$(PROJECT_NAME).xo $(build_strategy_path)
 	mkdir -p $(@D)
 	# Move into temp directory to avoid polluting top-level workdir
-	cd $(@D)/..; $(VPP) $(VPP_TARGET_SPECIFIC_ARGS) $(VPP_COMMON_ARGS) \
+	cd $(@D)/..; $(VPP) $(VPP_CONFIG_OVERLAYS) $(VPP_TARGET_SPECIFIC_ARGS) $(VPP_COMMON_ARGS) \
```
Comment (abejgonzalez) [inline platforms/vitis/cl_firesim/Makefile:149]: Shouldn't the default `VPP_TARGET_SPECIFIC_ARGS` be 1st so that the `build-bitstream.cfg` can be read in 1st?

**Review by abejgonzalez — CHANGES_REQUESTED** [review]

File: platforms/vitis/cl_firesim/Makefile (hunk 3 above)
Comment (davidbiancolin) [reply platforms/vitis/cl_firesim/Makefile:43]: It's useful for people like me that want to experiment on the configs. It doesn't make sense to me to shoe horn all of this behind the build strategies flag + this lets me point at files that might be elsewhere on the system.

Comment (davidbiancolin) [reply platforms/vitis/cl_firesim/Makefile:43]: Unless you mean why is this declaration needed? In that case, i think it just makes it more clear what the IF is.

File: platforms/vitis/cl_firesim/Makefile (hunk 2 above)
Comment (davidbiancolin) [reply platforms/vitis/cl_firesim/Makefile]: Done

File: platforms/vitis/cl_firesim/Makefile (hunk 4 above)
Comment (davidbiancolin) [reply platforms/vitis/cl_firesim/Makefile]: I wanted the overlays to be the highest precedence of the config files. The strategies should be non-overlapping with build-bitstream.cfg, but i want whatever to be passed in via the CONFIG_OVERLAY to win over everything except explicit commandline arguments

File: platforms/vitis/cl_firesim/Makefile (hunk 3 above)
Comment (abejgonzalez) [reply platforms/vitis/cl_firesim/Makefile:43]: The 1st response was sufficient. Thanks

**Review by abejgonzalez — APPROVED** [review]
LGTM
